test/lib/jdk/test/lib/process/Proc.java

```diff
@@ -282,18 +321,26 @@ public Proc start() throws IOException {
             }
         }
 
-        Collections.addAll(cmd, splitProperty("test.vm.opts"));
-        Collections.addAll(cmd, splitProperty("test.java.opts"));
-
-        if (cp == null) {
-            cmd.add("-cp");
-            cmd.add(System.getProperty("test.class.path") + File.pathSeparator +
-                    System.getProperty("test.src.path"));
-        } else if (!cp.isEmpty()) {
+        var lcp = fullcp();
+        if (lcp != null) {
             cmd.add("-cp");
-            cmd.add(cp.stream().collect(Collectors.joining(File.pathSeparator)));
+            cmd.add(lcp);
         }
 
+        if (compile) {
+            var comp = CompilerUtils.compile(
+                    Path.of(System.getProperty("test.src"), clazz + ".java"),
+                    Path.of(System.getProperty("test.classes")),
+                    cmd.subList(1, cmd.size()).toArray(new String[0]));
+                        // subList(1): all options added without launcher name
+            if (!comp) {
+                throw new IOException("Compilation error");
+            }
```

Comment:
Seems an odd use of IOException and not very helpful - will the actual compile command produce any diagnostics?

Reply:
Why not just RuntimeException()?